File: desktop-app/src/app/models/kanban/kanbanTask.ts
import { ShowUser } from "../user/show-user"
import { KanbanColumn } from "./kanbanColumn"

export class KanbanTaskChildrens {

    constructor(id: number, number: number, title: string, active: boolean) {
        this.id = id
        this.number = number
        this.title = title
        this.active = active
    }

    id: number
    number: number
    title: string
    active: boolean
}

export class KanbanTaskCreate {

    constructor(
        title: string,
        description: string,
        estimatedTime: number,
        kanbanColumnId: number,
        tags: string,
        importance: string,
        dueStartDate: Date,
        dueEndDate: Date,
        assignedUsernames: Array<string>,
        childrenIds: Array<number>
    ) {
        this.title = title
        this.description = description
        this.estimatedTime = estimatedTime
        this.kanbanColumnId = kanbanColumnId
        this.tags = tags
        this.importance = importance
        this.dueStartDate = dueStartDate
        this.dueEndDate = dueEndDate
        this.assignedUsernames = assignedUsernames
        this.childrenIds = childrenIds
    }

    title: string
    description: string
    estimatedTime: number
    kanbanColumnId: number
    tags: string
    importance: string
    dueStartDate: Date
    dueEndDate: Date
    assignedUsernames: Array<string>
    childrenIds: Array<number>
}

export class KanbanTaskUpdate {

    constructor(
        id: number,
        title: string,
        description: string,
        estimatedTime: number,
        tags: string,
        importance: string,
        dueStartDate: Date,
        dueEndDate: Date,
        assignedUsernames: Array<string>,
        childrenIds: Array<number>
    ) {
        this.id = id
        this.title = title
        this.description = description
        this.estimatedTime = estimatedTime
        this.tags = tags
        this.importance = importance
        this.dueStartDate = dueStartDate
        this.dueEndDate = dueEndDate
        this.assignedUsernames = assignedUsernames
        this.childrenIds = childrenIds
    }

    id: number
    title: string
    description: string
    estimatedTime: number
    tags: string
    importance: string
    dueStartDate: Date
    dueEndDate: Date
    assignedUsernames: Array<string>
    childrenIds: Array<number>
}

export class KanbanTaskMove {

    constructor(kanbanTaskId: number, newKanbanColumnId: number, newPosition: number) {
        this.kanbanTaskId = kanbanTaskId
        this.newKanbanColumnId = newKanbanColumnId
        this.newPosition = newPosition
    }

    kanbanTaskId: number
    newKanbanColumnId: number
    newPosition: number
}

export class KanbanTask {

    constructor(
        id: number,
        title: string,
        description: string,
        estimatedTime: number,
        creationDate: Date,
        number: number,
        orderInColumn: number,
        active: boolean,
        kanbanColumn: KanbanColumn,
        computedTime: number,
        creator: ShowUser,
        tags: string,
        importance: string,
        dueStartDate: Date,
        dueEndDate: Date,
        assignedUsers: Array<ShowUser>,
        children: Array<KanbanTask>
    ) {
        this.id = id
        this.title = title
        this.description = description
        this.estimatedTime = estimatedTime
        this.creationDate = creationDate
        this.number = number
        this.orderInColumn = orderInColumn
        this.active = active
        this.kanbanColumn = kanbanColumn
        this.computedTime = computedTime
        this.creator = creator
        this.tags = tags
        this.importance = importance
        this.dueStartDate = dueStartDate
        this.dueEndDate = dueEndDate
        this.assignedUsers = assignedUsers
        this.children = children
    }

    id: number
    title: string
    description: string
    estimatedTime: number
    creationDate: Date
    number: number
    orderInColumn: number
    active: boolean
    kanbanColumn: KanbanColumn
    computedTime: number
    creator: ShowUser
    tags: string
    importance: string
    dueStartDate: Date
    dueEndDate: Date
    assignedUsers: Array<ShowUser>
    children: Array<KanbanTask>
}

File: server-app-java/src/main/java/com/iman/model/kanban/KanbanTask.java
package com.iman.model.kanban;

import java.util.Date;
import java.util.List;

import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Index;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;
import javax.persistence.Table;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.PastOrPresent;

import org.hibernate.validator.constraints.Length;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIncludeProperties;
import com.iman.model.effort.Effort;
import com.iman.model.users.User;

import lombok.Data;

@Entity
@Data
@Table(name = "kanban_task", indexes = { @Index(columnList = "title"), @Index(columnList = "number"),
		@Index(columnList = "orderInColumn"), @Index(columnList = "active"), @Index(columnList = "creationDate"),
		@Index(columnList = "importance") })
public class KanbanTask {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	@NotBlank
	@Length(max = 50)
	private String title;

	@Length(max = 255)
	private String description;

	@Min(value = 0)
	private Double estimatedTime;

	@NotNull
	@PastOrPresent
	private Date creationDate;

	@NotNull
	@Min(value = 0)
	private Long number;

	@NotNull
	@Min(value = 0)
	private Long orderInColumn;

	private Boolean active;

	@ManyToOne
	@JoinColumn(name = "kanban_column_id")
	@JsonIgnore
	private KanbanColumn kanbanColumn;
	
	@OneToOne
	@JsonIncludeProperties({"id","username","imageUid"})
	private User creator;

	// EFFORTS

	@OneToMany(fetch = FetchType.LAZY, mappedBy = "kanbanTask")
	@JsonIgnore
	private List<Effort> efforts;

	public Double getComputedTime() {
		return efforts.stream().mapToDouble(Effort::getTime).sum();
	}

	// GANTT

	@Length(max = 50)
	private String importance;

	private Date dueStartDate;

	private Date dueEndDate;
	
	@ManyToMany
	@JoinTable(name = "kanban_task_assignation")
	@JsonIncludeProperties({"username", "imageUid"})
	private List<User> assignedUsers;
	
	@ManyToMany
	@JoinTable(name = "kanban_task_children")
	@JsonIncludeProperties({"id", "title", "number", "active"})
	private List<KanbanTask> children;

}
